presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/WeightedReservoirSampleStateStrategy.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.operator.aggregation.reservoirsample.WeightedDoubleReservoirSample;
+import com.facebook.presto.spi.PrestoException;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import static com.facebook.presto.operator.aggregation.differentialentropy.EntropyUtils.calculateEntropyFromSamples;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.String.format;
+
+public class WeightedReservoirSampleStateStrategy
+        implements DifferentialEntropyStateStrategy
+{
+    private final WeightedDoubleReservoirSample reservoir;
+
+    public WeightedReservoirSampleStateStrategy(long size)
+    {
+        if (size <= 0) {
```

Comment:
These checks should match the checks in WeightedDoubleReservoirSample's constructor; otherwise, the user will get an IllegalArgumentException when size is too large. I suggest to remove the checks here and update WeightedDoubleReservoirSample to throw PrestoException.

Reply:
Added checks here (the reservoir classes themselves seem too low level for that, IIUC).